### PR TITLE
Fcma two masks

### PR DESCRIPTION
**Pull request by yidawang**
Allow correlation-based voxel selection and training/classification to take two different masks, i.e. allow correlation computation takes place on two different components. The example and test codes are modified accordingly.

Thanks in advance for reviewing!

### CODE REVIEW

File: tests/fcma/test_io.py [hunk 1]
```diff
@@ -50,18 +57,30 @@ def test_prepare_fcma_data():
...
-            'raw data do not match'
+            'raw data do not match in test_prepare_mvpa_data'
     assert np.array_equal(labels, expected_labels), \
         'the labels do not match in test_prepare_mvpa_data'
 
+def test_prepare_searchlight_mvpa_data():
+    processed_data, labels = io.prepare_searchlight_mvpa_data(dir, extension, epoch_file)
+    expected_searchlight_processed_data = \
+        np.load(os.path.join(os.path.dirname(__file__),
+                             'data/expected_searchlight_processed_data.npy'))
```
Comment (mihaic) [inline tests/fcma/test_io.py]: If you use a `/` in your path string, there is little benefit from using `os.path.join`.

File: brainiak/fcma/classifier.py [hunk 2]
```diff
@@ -115,18 +122,24 @@ def __init__(self,
         self.num_digits_ = 0
         return
 
-    def _prepare_auto_corerelation_data(self, X,
-                                        start_voxel=0,
-                                        num_processed_voxels=None):
+    def _prepare_corerelation_data(self, x1, x2,
+                                   start_voxel=0,
+                                   num_processed_voxels=None):
         """ compute auto-correlation for the input data X
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: this is a leftover of the change from `X` to `x1` and `x2`. Fixed.

File: brainiak/fcma/classifier.py [hunk 3]
```diff
@@ -115,18 +122,24 @@ def __init__(self,
...
 
         Parameters
         ----------
-        X: a list of numpy array in shape [num_TRs, num_voxels]
-            len(X) is the number of samples
-            assuming all elements of X has the same num_voxels value
+        x1: a list of numpy array in shape [num_TRs, num_voxels]
+            x1 contains the activity data filtered by ROIs
+            and prepared for correlation computation.
+            Assuming all elements of x1 has the same num_voxels value,
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 4]
```diff
@@ -138,26 +151,30 @@ def _prepare_auto_corerelation_data(self, X,
...
         """
-        num_samples = len(X)
+        num_samples = len(x1)
         assert num_samples > 0, \
             'at least one sample is needed for correlation computation'
-        num_voxels = X[0].shape[1]
-        assert num_voxels == self.num_voxels_, \
-            'the number of voxels provided by X does not match ' \
-            'the number of voxels defined in the model'
+        num_voxels = x1[0].shape[1]
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 5]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
             else:
                 # do not store training data because it was partially computed
                 self.training_data_ = None
             logger.debug(
                 'kernel computation done'
             )
+        return data
+
+    def fit(self, X, y, num_training_samples=None):
+        """ use correlation data to train a model
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 6]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+        and then fit to a model defined by self.clf.
+
+        Parameters
+        ----------
+        X: tuple (x1, x2)
+            x1 and x2 are lists of numpy array in shape [num_TRs, num_voxels],
+            they contain the activity data filtered by ROIs
+            and prepared for correlation computation.
+            Assuming within list, all elements have the same num_voxels value
+        y: labels, 1D numpy array
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 7]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+            len(X) equals len(y)
+        num_training_samples: Optional[int]
+            Default None.
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 8]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+            The number of samples used in the training,
+            which is set when the kernel matrix is constructed
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 9]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+            portion by portion so the similarity vectors of the
+            test data have to be computed here.
+            This is ONLY set when sklearn.svm.SVC with
+            precomputed kernel is used.
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 10]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+            If it is set, only those samples will be used to fit the model.
+
+        Returns
+        -------
+        self: return the object itself
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: Done

File: brainiak/fcma/classifier.py [hunk 11]
```diff
@@ -350,63 +377,118 @@ def fit(self, X, y, num_training_samples=None):
...
+        """
+        time1 = time.time()
+        # if X is not a tuple, make it one but replicating it
+        if not isinstance(X, tuple):
+            X = (X, X)
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: This allows users to input a single `X`. Let me explain this in the docstring instead of removing it.

File: brainiak/fcma/cython_blas.pyx [hunk 12]
```diff
@@ -65,6 +66,10 @@ def compute_self_corr_for_voxel_sel(py_trans_a, py_trans_b, py_m, py_n, py_k,
     the starting voxel of assigned voxels
     used to locate the second matrix B
 
+    py_b: 2D array in shape [epoch_length, num_voxels] in our case
```
Comment (yidawang) [inline brainiak/fcma/cython_blas.pyx]: done

File: brainiak/fcma/io.py [hunk 13]
```diff
@@ -32,7 +32,7 @@
 logger = logging.getLogger(__name__)
 
 
-def _read_one_nifti_file(filename, mask=None):
+def _read_one_nifti_file(filename, mask=None, mask2=None):
```
Comment (yidawang) [inline brainiak/fcma/io.py]: done

File: brainiak/fcma/io.py [hunk 14]
```diff
@@ -32,7 +32,7 @@
...
     """ read data from an NIfTI file, apply the mask if provided
```
Comment (yidawang) [inline brainiak/fcma/io.py]: done

File: brainiak/fcma/io.py [hunk 15]
```diff
@@ -43,27 +43,42 @@ def _read_one_nifti_file(filename, mask=None):
         This is a 3D binary file indicating which voxels
         are selected (labelled 1).
         If it is not specified, the data will not be masked and remain in 4D
+    mask2: Optional[3D array]
+        This is a 3D binary file indicating which voxels
+        are selected (labelled 1).
```
Comment (yidawang) [inline brainiak/fcma/io.py]: done

File: brainiak/fcma/io.py [hunk 16]
```diff
@@ -43,27 +43,42 @@ def _read_one_nifti_file(filename, mask=None):
...
+        It only takes effect when mask is not None.
```
Comment (yidawang) [inline brainiak/fcma/io.py]: done

File: brainiak/fcma/io.py [hunk 17]
```diff
@@ -43,27 +43,42 @@ def _read_one_nifti_file(filename, mask=None):
...
+        If it is not specified, the returning variable data2 is None
 
     Returns
     -------
     data: 2D/4D array
```
Comment (yidawang) [inline brainiak/fcma/io.py]: done

File: brainiak/fcma/voxelselector.py [hunk 18]
```diff
@@ -42,27 +42,32 @@ class VoxelSelector:
...
                     2. the activity data has been z-scored,
                        ready to compute correlation as matrix multiplication
                     3. all subjects have the same number of epochs
                     4. epochs belonging to the same subject are adjacent
                        in the list
-                    5. voxel selection is always done in the auto-correlation,
-                       i.e. raw_data correlate with themselves
-                    6. if MPI jobs are running on multiple nodes, the user
+                    5. if MPI jobs are running on multiple nodes, the user
                        home directory is shared by all nodes
```
Comment (yidawang) [inline brainiak/fcma/voxelselector.py]: done

File: tests/fcma/test_io.py (hunk 1 above)
Comment (yidawang) [inline tests/fcma/test_io.py]: it is a two-level file layout. I don't want the data files to be in the same level of code files, so create a directory `data` to host all the data files.

File: brainiak/fcma/classifier.py [hunk 19]
```diff
@@ -115,18 +122,24 @@ def __init__(self,
         self.num_digits_ = 0
         return
 
-    def _prepare_auto_corerelation_data(self, X,
-                                        start_voxel=0,
-                                        num_processed_voxels=None):
+    def _prepare_corerelation_data(self, x1, x2,
```
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 20]
```diff
@@ -149,31 +149,31 @@ def _prepare_corerelation_data(self, x1, x2,
...
+        num_samples = len(X1)
         assert num_samples > 0, \
             'at least one sample is needed for correlation computation'
-        num_voxels = x1[0].shape[1]
-        num_voxels2 = x2[0].shape[1]
-        assert num_voxels*num_voxels2 == self.num_features_, \
+        num_voxels1 = X1[0].shape[1]
+        num_voxels2 = X2[0].shape[1]
+        assert num_voxels1 * num_voxels2 == self.num_features_, \
             'the number of features provided by X does not match ' \
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: `X1`?

File: brainiak/fcma/classifier.py [hunk 21]
```diff
@@ -149,31 +149,31 @@ def _prepare_corerelation_data(self, x1, x2,
...
             'the number of features defined in the model'
-        assert x1[0].shape[0] == x2[0].shape[0], \
+        assert X1[0].shape[0] == X2[0].shape[0], \
             'the numbers of TRs of X and X2 are not identical'
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: "X1 and X2"

File: brainiak/fcma/classifier.py [hunk 22]
```diff
@@ -289,15 +289,15 @@ def _compute_kernel_matrix_in_portion(self, x1, x2):
 
         Parameters
         ----------
-        x1: a list of numpy array in shape [num_TRs, num_voxels]
+        X1: a list of numpy array in shape [num_TRs, num_voxels]
             X contains the activity data filtered by ROIs
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: "X1"

File: brainiak/fcma/classifier.py [hunk 23]
```diff
@@ -356,15 +356,15 @@ def _generate_training_data(self, x1, x2, num_training_samples):
...
-        x2: a list of numpy array in shape [num_TRs, num_voxels]
-            len(x1) equals len(x2).
-            Assuming all elements of x2 has the same num_voxels value.
-            x2 can be identical to x1; if not, x always has more voxels
-            than x2.
+            All elements of X1 must have the same num_voxels value,
+        X2: a list of numpy array in shape [num_TRs, num_voxels]
+            len(X1) equals len(X2).
+            All elements of X2 must have the same num_voxels value.
+            X2 can be identical to X1; if not, x always has more voxels
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: "X1 must have more voxels..."

File: brainiak/fcma/cython_blas.pyx [hunk 24]
```diff
@@ -55,7 +55,8 @@ def compute_self_corr_for_voxel_sel(py_trans_a, py_trans_b, py_m, py_n, py_k,
     py_alpha: float
     the weight applied to the first matrix A
 
-    py_a: 2D array in shape [epoch_length, num_voxels] in our case
+    py_a: 2D array
+    In shape [epoch_length, num_voxels] in our case, it is
     the activity data of an epoch, part 1 of the data to be correlated with.
```
Comment (mihaic) [inline brainiak/fcma/cython_blas.pyx]: This still does not read like a sentence:
"In shape [epoch_length, num_voxels] in our case, it is the activity data of an epoch, part 1 of the data to be correlated with."
Also, we usually put the shape in the type line.

Comment (yidawang) [inline brainiak/fcma/cython_blas.pyx]: fixed

File: brainiak/fcma/classifier.py (hunk 20 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: It is `X` (correlation of `X1` and `X2`), let me make it clearer.

File: brainiak/fcma/classifier.py (hunk 21 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: fixed

File: brainiak/fcma/classifier.py (hunk 22 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py (hunk 23 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: done

File: brainiak/fcma/classifier.py [hunk 25]
```diff
@@ -433,15 +433,13 @@ def fit(self, X, y, num_training_samples=None):
 
         Parameters
         ----------
-        X: tuple (X1, X2)
-            X1 and X2 are lists of numpy array in shape [num_TRs, num_voxels],
-            they contain the activity data filtered by ROIs
+        X: list of tuple (x1, x2)
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: Capitals, please, also elsewhere.

File: brainiak/fcma/classifier.py [hunk 26]
```diff
@@ -459,16 +457,12 @@ def fit(self, X, y, num_training_samples=None):
...
-        assert len(X1) == len(y), \
-            'the number of samples does not match the number labels'
-        assert len(X1) == len(X2), \
-            'the numbers of samples in X1 and X2 are not identical'
+        assert len(X) == len(y), \
+            'the number of samples must be equal to the number of labels'
+        for x in X:
+            assert len(x) == 2, \
+                'there must be two parts for each correlation computation'
+        X1, X2 = zip(*X)
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: Note that `X1` and `X2` will be iterators. In case you want them to be actual lists later, you must create lists based on them.

File: brainiak/fcma/classifier.py (hunk 5 above)
Comment (mihaic) [inline brainiak/fcma/classifier.py]: Actually, you missed the first line:
`"""Use correlation data to train a model.`

File: brainiak/fcma/classifier.py [hunk 27]
```diff
@@ -350,63 +377,115 @@ def fit(self, X, y, num_training_samples=None):
...
+        """ use correlation data to train a model
+
+        First compute the correlation of the input data,
+        and then normalize within subject
+        if more than one sample in one subject,
+        and then fit to a model defined by self.clf.
+
+        Parameters
+        ----------
+        X: list of tuple (x1, x2)
```
Comment (mihaic) [inline brainiak/fcma/classifier.py]: `X1` and `X2` are still multidimensional arrays, so should be capitalized everywhere.

File: brainiak/fcma/classifier.py (hunk 25 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: `x1` and `x2` are elements of the lists `X1` and `X2`.

File: brainiak/fcma/classifier.py (hunk 26 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: "zip() in conjunction with the * operator can be used to unzip a list"
from https://docs.python.org/3/library/functions.html#zip

File: brainiak/fcma/classifier.py (hunk 27 above)
Comment (yidawang) [inline brainiak/fcma/classifier.py]: I understand in this way: `X = [(x1, x2), (x1, x2), ..., (x1, x2)]`, `X1=[x1, x1, ..., x1]`
